lib/src/strings.dart

```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
-  String result;
-
-  final trimmedString = string.trimRight();
-  final removedPortion = string.substring(trimmedString.length);
+String _tryYamlEncodeFolded(String string, int indentSize, String lineEnding) {
+  final indent = ' ' * indentSize;
+
+  var (literalPrefix, trimmedString, stripped) = _prepareBlockString(
+    '>',
+    string,
+    indent,
+    lineEnding,
+  );
+
+  // Try folding if string has any non-empty values
+  if (trimmedString.isNotEmpty) {
+    final trimmedSplit = trimmedString.split(lineEnding);
+
+    // Apply the left indent. First line didn't have indent.
+    final first = indent + trimmedSplit[0];
+
+    if (trimmedSplit.length > 1) {
+      /// Try folding to match specification:
+      /// * https://yaml.org/spec/1.2.2/#65-line-folding
+      trimmedString =
+          trimmedSplit.foldIndexed(first, (index, previous, current) {
+        if (index == 0) return previous;
+
+        /// If initially empty, this line holds only `\n`. It's okay. Also, if
+        /// this line is just an empty space.
+        ///
+        /// See https://yaml.org/spec/1.2.2/#64-empty-lines
+        if (current.trim().isEmpty) return previous + lineEnding + current;
+
+        var updated = current;
+
+        /// For consecutive non-empty lines, we add a `\n` with a space.
+        ///
+        /// We need to ensure the ACTUAL previous string was non-empty
+        if (trimmedSplit[index - 1].trim().isNotEmpty) {
+          /// We remove indent and check if the string starts with a space.
+          ///
+          /// Apply `\n` for `foo\nbar` but not `foo\n bar`.
```

Comment:
I'm not sure follow 100%, so you're trying to:
 * Insert a line break into `foo\nbar` -> `foo\n\nbar`,
 * But not do that for `foo\n bar`.

With the added complexity that we have indentation, so it's actually `${indent}foo\n${indent}bar` and not `foo\nbar`.

Might I suggest that duplicating the logic in `_prepareBlockString` could make this simpler.
What if we just did `string.trimRigh().split('\n').reduce(...` or something like it.

It probably makes sense to do this along with the change where `_tryYamlEncodeFolded` can return `null`, if the value can't be encoded as a folded string.

Reply:
Yeah. `Yaml` uses the first line for indent information. So for `foo\n bar`:
1. Encode as `>-${indent}foo\n${indent} bar`.
2. When parsed, the indent on `foo` allows us to have the space before `bar`

For `foo\nbar`:
1. Encode as `>-${indent}foo\n\n${indent}bar`.
2. When parsed, [yaml] gets rid of the first line break since the next one is also an empty line that giving us `foo\nbar`

[yaml]: https://yaml.org/spec/1.2.2/#65-line-folding:~:text=Line%20folding%20allows%20long%20lines%20to%20be%20broken%20for%20readability%2C%20while%20retaining%20the%20semantics%20of%20the%20original%20long%20line.%20If%20a%20line%20break%20is%20followed%20by%20an%20empty%20line%2C%20it%20is%20trimmed%3B%20the%20first%20line%20break%20is%20discarded%20and%20the%20rest%20are%20retained%20as%20content.